sql/report_util_functions.sql

```diff
@@ -1483,4 +1483,295 @@ BEGIN
+    p_extendedMonths INT(11)) RETURNS TINYINT(1)
+    DETERMINISTIC
+BEGIN
+    DECLARE result TINYINT(1) DEFAULT 0;
+    DECLARE uuidHIVTestDate VARCHAR(38) DEFAULT "c6c08cdc-18dc-4f42-809c-959621bc9a6c";
+
+    SELECT
+        TRUE INTO result
+    FROM obs o
+    JOIN concept c ON c.concept_id = o.concept_id AND c.retired = 0
+    WHERE o.voided = 0
+        AND o.person_id = p_patientId
+        AND c.uuid = uuidHIVTestDate
+        AND o.value_datetime IS NOT NULL
+        AND timestampdiff(MONTH, CAST(o.value_datetime AS DATE), p_endDate) > p_extendedMonths
+        LIMIT 1;
+
+    RETURN (result );
+END$$
+DELIMITER ;
+
+-- patientIsFemale
+
+DROP FUNCTION IF EXISTS patientIsFemale;
+
+DELIMITER $$
+CREATE FUNCTION patientIsFemale(
+    p_patientId INT(11),
+    p_gender VARCHAR(1)) RETURNS TINYINT(1)
+    DETERMINISTIC
+BEGIN 
+    DECLARE patientIsFemale TINYINT(1) DEFAULT 0;
+
+    SELECT TRUE INTO patientIsFemale
+    FROM person p
+    WHERE p.person_id = p_patientId
+        AND p.gender = p_gender;
+        
+    RETURN (patientIsFemale );
+END$$
+DELIMITER ;
+
+-- hivStatusKnown3MonthsOrLessBeforeReportEndDate
+
+DROP FUNCTION IF EXISTS hivStatusKnown3MonthsOrLessBeforeReportEndDate;
+
+DELIMITER $$
+CREATE FUNCTION hivStatusKnown3MonthsOrLessBeforeReportEndDate(
+    p_patientId INT(11),
+    p_endDate DATE,
+    p_extendedMonths INT(11)) RETURNS TINYINT(1)
+    DETERMINISTIC
+BEGIN
+    DECLARE result TINYINT(1) DEFAULT 0;
+    DECLARE uuidHIVTestDate VARCHAR(38) DEFAULT "c6c08cdc-18dc-4f42-809c-959621bc9a6c";
+    
+    SELECT
+        TRUE INTO result
+    FROM obs o
+    JOIN concept c ON c.concept_id = o.concept_id AND c.retired = 0
+    WHERE o.voided = 0
+        AND o.person_id = p_patientId
+        AND c.uuid = uuidHIVTestDate
+        AND o.value_datetime IS NOT NULL
+        AND timestampdiff(MONTH, CAST(o.value_datetime AS DATE), p_endDate) <= p_extendedMonths
+        LIMIT 1;
+
+    RETURN (result );
+END$$
+DELIMITER ;
+
+-- patientHIVRetestResultIsPositive
+
+DROP FUNCTION IF EXISTS patientHIVRetestResultIsPositive;
+
+DELIMITER $$
+CREATE FUNCTION patientHIVRetestResultIsPositive(
+    p_patientId INT(11)) RETURNS TINYINT(1)
+    DETERMINISTIC
+BEGIN
+    DECLARE result TINYINT(1) DEFAULT 0;
+    DECLARE uuidRepeatTestResult VARCHAR(38) DEFAULT "7682c09b-8e81-4e30-8afd-636fb9fcd4a1";
+    DECLARE uuidRepeatTestResultIsPositive VARCHAR(38) DEFAULT "7acfafa4-f19b-485e-97a7-c9e002dbe37a";
+    SELECT
+        TRUE INTO result
+    FROM obs o
+    JOIN concept c ON c.concept_id = o.concept_id AND c.retired = 0
+    WHERE o.voided = 0
+        AND o.person_id = p_patientId
+        AND c.uuid = uuidRepeatTestResult
+        AND o.value_coded IS NOT NULL
+        AND o.value_coded = (SELECT concept_id FROM concept WHERE uuid = uuidRepeatTestResultIsPositive)
+        LIMIT 1;
+
+    RETURN (result );
+END$$
+DELIMITER ;
+
+-- patientStatusIsAlreadyOnART
+
+DROP FUNCTION IF EXISTS patientStatusIsAlreadyOnART;
```

Comment:
@prashantsable1991 Why are these 2 functions created (patientStatusIsAlreadyOnART and patientStatusIsNotAlreadyOnART), isn't it enough to have one?

Reply:
```OR
        (
            hivStatusKnownMoreThan3MonthsBeforeReportEndDate(pat.patient_id, p_endDate, 3) AND
            NOT patientStatusIsAlreadyOnART(pat.patient_id)
        )```